src/service/gateway.rs

```diff
@@ -0,0 +1,278 @@
+                let GatewayScIsActiveRespV1 {
+                    sc_id, sc_owner, ..
+                } = &resp;
+                if sc_id == id && sc_owner == owner {
+                    Ok(resp)
+                } else {
+                    Err(Error::custom("mismatched state channel id and owner"))
+                }
+            }
+            Some(other) => Err(Error::custom(format!(
+                "invalid is_active response {other:?}",
+            ))),
+            None => Err(Error::custom("empty is_active response")),
+        }
+    }
+
+    async fn get_config(&mut self, keys: Vec<String>) -> Result<GatewayRespV1> {
+        let resp = self
+            .client
+            .config(GatewayConfigReqV1 { keys })
+            .await?
+            .into_inner();
+        resp.verify(&self.uri.pubkey)?;
+        Ok(resp)
+    }
+
+    pub async fn config(&mut self, keys: Vec<String>) -> Result<Vec<BlockchainVarV1>> {
+        match self.get_config(keys).await?.msg {
+            Some(gateway_resp_v1::Msg::ConfigResp(GatewayConfigRespV1 { result })) => Ok(result),
+            Some(other) => Err(Error::custom(format!("invalid config response {other:?}"))),
+            None => Err(Error::custom("empty config response")),
+        }
+    }
+
+    pub async fn height(&mut self) -> Result<(u64, u64)> {
+        let resp = self.get_config(vec![]).await?;
+        Ok((resp.height, resp.block_age))
+    }
+
+    pub async fn validators(&mut self, quantity: u32) -> Result<Vec<KeyedUri>> {
+        let resp = self
+            .client
+            .validators(GatewayValidatorsReqV1 { quantity })
+            .await?
+            .into_inner();
+        resp.verify(&self.uri.pubkey)?;
+        match resp.msg {
+            Some(gateway_resp_v1::Msg::ValidatorsResp(GatewayValidatorsRespV1 { result })) => {
+                result.into_iter().map(KeyedUri::try_from).collect()
+            }
+            Some(other) => Err(Error::custom(format!(
+                "invalid validator response {other:?}"
+            ))),
+            None => Err(Error::custom("empty validator response")),
+        }
+    }
+
+    pub async fn version(&mut self) -> Result<u64> {
+        let resp = self
+            .client
+            .version(GatewayVersionReqV1 {})
+            .await?
+            .into_inner();
+        resp.verify(&self.uri.pubkey)?;
+        match resp.msg {
+            Some(gateway_resp_v1::Msg::Version(GatewayVersionRespV1 { version })) => Ok(version),
+            Some(other) => Err(Error::custom(format!(
+                "invalid validator response {other:?}"
+            ))),
+            None => Err(Error::custom("empty version response")),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct GatewayVersion {
+    major: u16,
+    minor: u16,
+    patch: u16,
+}
+
+impl From<u64> for GatewayVersion {
+    fn from(v: u64) -> Self {
+        let patch = (v % 10000) as u16;
```

Comment:
is it wishful thinking to ask for a simpler versioning scheme? Like each byte be a single field?

Reply:
Erlang miner compatibility.. only there until validators go away.